### PR TITLE
python3Packages.parse: 1.16.0 -> 1.18.0, python-docx: run behave tests

### PR DESCRIPTION
**Pull request by maxxk**
###### Motivation for this change

Fix `behave` which was broken from previous `parse` update (see `nixpkgs-review` from #95690; https://github.com/behave/behave/issues/862).

ZHF: #97479

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

nixpkgs-review:

```
3 packages failed to build:
python37Packages.pytest-bdd python38Packages.pytest-bdd terraform-compliance

17 packages built:
nrfutil python27Packages.parse python27Packages.parse-type python27Packages.pyparser python27Packages.pytest-bdd python37Packages.behave python37Packages.parse python37Packages.parse-type python37Packages.pyparser python37Packages.python-docx python37Packages.radish-bdd python38Packages.behave python38Packages.parse python38Packages.parse-type python38Packages.pyparser python38Packages.python-docx python38Packages.radish-bdd
```

Failed packages also fail on master.

### CODE REVIEW

**Comment by risicle** [comment]
`behave` fails to build for me on macos 10.14:

```
=================================== FAILURES ===================================
___ TestAsyncStepDecoratorPy34.test_step_decorator_async_run_until_complete2 ___

self = <tests.api._test_async_step34.TestAsyncStepDecoratorPy34 object at 0x1053c5150>

    def test_step_decorator_async_run_until_complete2(self):
        step_container = SimpleStepContainer()
        with use_step_import_modules(step_container):
            # -- STEP-DEFINITIONS EXAMPLE (as MODULE SNIPPET):
            # VARIANT 2: Use @asyncio.coroutine def step_impl()
            from behave import step
            from behave.api.async_step import async_run_until_complete
            import asyncio
    
            @step('a tagged-coroutine async step waits "{duration:f}" seconds')
            @async_run_until_complete
            @asyncio.coroutine
            def step_async_step_waits_seconds2(context, duration):
                yield from asyncio.sleep(duration)
    
        # -- USES: async def step_impl(...) as async-step (coroutine)
        AsyncStepTheory.validate(step_async_step_waits_seconds2)
    
        # -- RUN ASYNC-STEP: Verify that it is behaving correctly.
        # ENSURE: Execution of async-step matches expected duration.
        context = Context(runner=Runner(config={}))
        with StopWatch() as stop_watch:
            step_async_step_waits_seconds2(context, duration=0.2)
>       assert abs(stop_watch.duration - 0.2) <= 0.05
E       AssertionError

tests/api/_test_async_step34.py:75: AssertionError
___ TestAsyncStepDecoratorPy35.test_step_decorator_async_run_until_complete1 ___

self = <tests.api._test_async_step35.TestAsyncStepDecoratorPy35 object at 0x1051bd890>

    def test_step_decorator_async_run_until_complete1(self):
        step_container = SimpleStepContainer()
        with use_step_import_modules(step_container):
            # -- STEP-DEFINITIONS EXAMPLE (as MODULE SNIPPET):
            # VARIANT 1: Use async def step_impl()
            from behave import step
            from behave.api.async_step import async_run_until_complete
            import asyncio
    
            @step('an async coroutine step waits "{duration:f}" seconds')
            @async_run_until_complete
            async def step_async_step_waits_seconds(context, duration):
                await asyncio.sleep(duration)
    
        # -- USES: async def step_impl(...) as async-step (coroutine)
        AsyncStepTheory.validate(step_async_step_waits_seconds)
    
        # -- RUN ASYNC-STEP: Verify that it is behaving correctly.
        # ENSURE: Execution of async-step matches expected duration.
        context = Context(runner=Runner(config={}))
        with StopWatch() as stop_watch:
            step_async_step_waits_seconds(context, 0.2)
>       assert abs(stop_watch.duration - 0.2) <= 0.05
E       AssertionError

tests/api/_test_async_step35.py:75: AssertionError
=============================== warnings summary ===============================
tests/unit/test_behave4cmd_command_shell_proc.py:26
  /private/tmp/nix-build-behave-1.2.7.dev1.drv-1/source/tests/unit/test_behave4cmd_command_shell_proc.py:26: DeprecationWarning: invalid escape sequence \w
    winpath_pattern = u"^([A-Za-z]:(\\[\w\.\-]+)+)|((\\[\w\.\-]+)*)|(\s[\w\.\-]+([\w\.\-]+)*)$"

-- Docs: https://docs.pytest.org/en/latest/warnings.html
- generated xml file: /private/tmp/nix-build-behave-1.2.7.dev1.drv-1/source/build/testing/report.xml -
- generated html file: file:///private/tmp/nix-build-behave-1.2.7.dev1.drv-1/source/build/testing/report.html -
=========================== short test summary info ============================
FAILED tests/api/test_async_step.py::TestAsyncStepDecoratorPy34::test_step_decorator_async_run_until_complete2
FAILED tests/api/test_async_step.py::TestAsyncStepDecoratorPy35::test_step_decorator_async_run_until_complete1
======= 2 failed, 1248 passed, 7 skipped, 2 xfailed, 1 warning in 12.80s =======
```

This looks like its a timing-based test, which are always trouble - perhaps it should be disabled?

**Review by drewrisinger — APPROVED** [review]
* [x] Diff LGTM
* [x] Commits LGTM
* [x] Confirm nix-review results from author:

```
https://github.com/NixOS/nixpkgs/pull/97737
2 packages failed to build:
python38Packages.pytest-bdd terraform-compliance

18 packages built:
nrfutil python27Packages.parse python27Packages.parse-type python27Packages.pyparser python27Packages.pytest-bdd python37Packages.behave python37Packages.parse python37Packages.parse-type python37Packages.pyparser python37Packages.pytest-bdd python37Packages.python-docx python37Packages.radish-bdd python38Packages.behave python38Packages.parse python38Packages.parse-type python38Packages.pyparser python38Packages.python-docx python38Packages.radish-bdd
```

**Comment by maxxk** [comment]
> `behave` fails to build for me on macos 10.14:

@risicle I tried several times building both `python38Packages.behave` and, as in your example, `python37Packages.behave` but unfortunately I couldn't reproduce the issue using fresh Nix install on macOS 10.14.6 18G6020 :( 

Probably we could ping someone maintaining nixpkgs on darwin to continue investigating?

In the meantime, could you please check that changing `pytest` command to the following fixes the build for you:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/behave/default.nix#L33
```
pytest tests -k "not test_step_decorator_async_run_until_complete"
```

**Comment by risicle** [comment]
I simply think I've got slower computers than you (maybe try building on a heavily loaded machine), and as I say, timing-based tests are just trouble sooner or later. It's not too much of a loss disabling just `test_step_decorator_async_run_until_complete*` is it?

**Comment by drewrisinger** [comment]
You can disable only that test if using ``pytestCheckHook`` and ``disableTests``

```nix
checkInputs = [ pytestCheckHook ... ];

disabledTests = lib.optionals stdenv.isDarwin [ "test_step_decorator_async_run_until_complete" ];
```

**Comment by maxxk** [comment]
Thanks, I added the change. I'd just prefer to fix the test, at least by increasing the timeout, but skipping it on darwin is fine too.

File: pkgs/development/python-modules/behave/default.nix [hunk 1]
```diff
@@ -16,22 +16,21 @@ buildPythonApplication rec {
...
+  checkInputs = [ mock pathpy pyhamcrest pytest pytest-html pytestCheckHook ];
   buildInputs = [ glibcLocales ];
   propagatedBuildInputs = [ colorama cucumber-tag-expressions parse parse-type six ];
 
   postPatch = ''
     patchShebangs bin
   '';
 
-  doCheck = true;
+  pytestFlagsArray = [ "tests/" ];
```
Comment (drewrisinger) [inline pkgs/development/python-modules/behave/default.nix:27]: Probably redundant b/c pytest would automatically find that folder, but not important.

File: pkgs/development/python-modules/behave/default.nix [hunk 2]
```diff
@@ -16,22 +16,20 @@ buildPythonApplication rec {
...
+  checkInputs = [ mock pathpy pyhamcrest pytest pytest-html pytestCheckHook ];
   buildInputs = [ glibcLocales ];
   propagatedBuildInputs = [ colorama cucumber-tag-expressions parse parse-type six ];
 
   postPatch = ''
     patchShebangs bin
   '';
 
-  doCheck = true;
+  disabledTests = stdenv.lib.optionals stdenv.isDarwin [ "test_step_decorator_async_run_until_complete" ];
```
Comment (jonringer) [inline pkgs/development/python-modules/behave/default.nix:27]: do you mind adding a comment as to why disabling this test is necessary?

File: pkgs/development/python-modules/behave/default.nix [hunk 3]
```diff
@@ -16,22 +16,22 @@ buildPythonApplication rec {
     sha256 = "1ssgixmqlg8sxsyalr83a1970njc2wg3zl8idsmxnsljwacv7qwv";
   };
 
-  checkInputs = [ mock pathpy pyhamcrest pytest pytest-html ];
+  checkInputs = [ mock pathpy pyhamcrest pytest pytest-html pytestCheckHook ];
```
Comment (drewrisinger) [inline pkgs/development/python-modules/behave/default.nix:19]:
```suggestion
  checkInputs = [ pytestCheckHook mock pathpy pyhamcrest pytest-html ];
```
``pytest`` is redundant with ``pytestCheckHook``. Personal preference is to put ``pytestCheckHook`` first b/c it's the test runner.

File: pkgs/development/python-modules/behave/default.nix [hunk 4]
```diff
@@ -1,6 +1,6 @@
 { stdenv, fetchFromGitHub
 , buildPythonApplication, python
-, mock, pathpy, pyhamcrest, pytest, pytest-html
+, mock, pathpy, pyhamcrest, pytest, pytest-html, pytestCheckHook
```
Comment (drewrisinger) [inline pkgs/development/python-modules/behave/default.nix:3]:
```suggestion
, pytestCheckHook, mock, pathpy, pyhamcrest, pytest-html,
```

**Review by drewrisinger — CHANGES_REQUESTED** [review]
* [ ] Minor point in diff still: redundant ``pytest``
* [x] Commits LGTM
* [x] Builds via ``nix-review``: Failed packages fail on ``master``

```
https://github.com/NixOS/nixpkgs/pull/97737
2 packages failed to build:
python38Packages.pytest-bdd terraform-compliance

18 packages built:
nrfutil python27Packages.parse python27Packages.parse-type python27Packages.pyparser python27Packages.pytest-bdd python37Packages.behave python37Packages.parse python37Packages.parse-type python37Packages.pyparser python37Packages.pytest-bdd python37Packages.python-docx python37Packages.radish-bdd python38Packages.behave python38Packages.parse python38Packages.parse-type python38Packages.pyparser python38Packages.python-docx python38Packages.radish-bdd
```

**Comment by jonringer** [comment]
sorry @maxxk that this turned out to be such a ..... pain of a PR

but thanks for sticking through it :)

**Review by jonringer — APPROVED** [review]
LGTM

failures are broken on target branch
```
https://github.com/NixOS/nixpkgs/pull/97737
3 packages failed to build:
python37Packages.pytest-bdd python38Packages.pytest-bdd terraform-compliance

17 packages built:
nrfutil python27Packages.parse python27Packages.parse-type python27Packages.pyparser python27Packages.pytest-bdd python37Packages.behave python37Packages.parse python37Packages.parse-type python37Packages.pyparser python37Packages.python-docx python37Packages.radish-bdd python38Packages.behave python38Packages.parse python38Packages.parse-type python38Packages.pyparser python38Packages.python-docx python38Packages.radish-bdd
```

**Pull request merged**